christfellowshipchurch/web-app
# [Events] Cleanup EventsPanel and overhaul styling

# Changes or Updates
- Adds `styled-components` 💅
- Creates [src/styles/theme.js](https://github.com/christfellowshipchurch/web-app/pull/158/files#diff-8d0114a2287c73495fa4178acf42f60f) to mimic/copy core Bootstrap color and spacing values
- Adds an `<EventPanel>` component with tab switching mechanism, and refactors where/how `<EventChat>` is used
- Adds `ChatConversation` icon
- Prevent guests from chatting by showing a log in prompt instead of message input
- Create a new `ui/chat` folder to organize smaller custom components for chat
- Use new custom `<Message>` component for closer CF branding
- Use `<LiveIndicator>` component that was [refactored in](https://github.com/christfellowshipchurch/web-app/pull/152/files#diff-0ad2d11fed67ec0feb40143a16534f84) current/"old" version of **Events** page
- Better loading states for chat panel
- Export `Icon` from `ui` root index for convenience

### Styled Components 💅
Added so we can begin exploring a non-Bootstrapped based development workflow. This implementation coexists with Bootstrap, and shouldn't cause any global conflicts.

For the transition/interim where `styled-components` and Bootstrap co-exist, the _src/styles/theme.js_ file provides important values from the app's CSS context to it's JS one. See [Styled Components > Theming](https://styled-components.com/docs/advanced#theming) for details.

A React component will typically access the theme in one of two ways:

#### Directly in style definition:
```js
const PanelHeader = styled.div`
  display: flex;
  justify-content: space-between;
  background: ${({ theme }) => theme.card.background}; /* 💅  Theme value */
`;
```

#### via `withTheme` HOC
```jsx
const ThemedTab = withTheme(({ theme, label, active, onPress }) => (
  <TabContainer>
    <TabButton active={active} onClick={onPress}>
      <Icon fill={active ? theme.brand : theme.card.color} name="home" size={36} /> {/* 💅 Theme value */ }
      <TabLabel>
        {label}
      </TabLabel>
    </TabButton>
  </TabContainer>
));
```

## Tradeoffs/Caveats
Still have not meaningfully addressed responsive web layout/mobile functionality.

## Screenshots

|Overall Layout with Event Panel|
|---|
|<img src="https://user-images.githubusercontent.com/1812955/93166271-1ef04900-f6ec-11ea-835b-9b63fcda46c4.jpg" width="480" />|

|Logged Out (Guest)|Logged In|Switching Between|
|---|---|---|
|<img src="https://user-images.githubusercontent.com/1812955/93166262-1ac42b80-f6ec-11ea-8310-965445f72985.jpg" width="350"/>|<img src="https://user-images.githubusercontent.com/1812955/93166263-1bf55880-f6ec-11ea-8793-7cf4a7e71943.jpg" width="350"/>|![2020-09-14 13 28 48](https://user-images.githubusercontent.com/1812955/93166387-637be480-f6ec-11ea-8131-307c8146211e.gif)|

|User Profile Image States|
|---|
|<img src="https://user-images.githubusercontent.com/1812955/93166414-6d9de300-f6ec-11ea-8378-eb25f0defffd.jpg" width="480" />|





# Test Instructions
#### ⚠️ Note on `/etc/hosts`
The app's fonts package host only serves requests from *.christfellowship.church subdomains. This means viewing the site locally view http://localhost:3000 relies on fallback system sans-serif fonts.
To workaround, I suggest adding an entry to your `/etc/hosts` file like so:
```
127.0.0.1 local.christfellowship.church
```
No, you can go to http://local.christfellowship.church:3000 and see the local site _with_ proper font packages.

### Test Flows
1. Open an event locally 👉 http://local.christfellowship.church:3000/events/test-24-hour-live-stream
1. Verify there is now a panel to the right of the video with tabs for Chat and Schedule
1. While logged out, verify you cannot send messages in the chat, and are prompted to log in instead
1. Verify you can log in view the prompt at the bottom of chat
1. While logged in, verify chat reloads properly and you can send messages as yourself

# Tests
